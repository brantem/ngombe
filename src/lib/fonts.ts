import { Nunito } from 'next/font/google';

export const nunito = Nunito({ subsets: ['latin'], weight: ['500', '800'] });
