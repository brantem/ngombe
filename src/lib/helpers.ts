import { twMerge } from 'tailwind-merge';
import clsx, { ClassValue } from 'clsx';

/* c8 ignore next 3 */
export const cn = (...inputs: ClassValue[]) => {
  return twMerge(clsx(inputs));
};

export const calcPercentage = (a: number, b: number) => {
  return Math.round((a / b) * 100);
};
