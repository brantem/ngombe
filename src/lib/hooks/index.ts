import { useState, useLayoutEffect, useEffect } from 'react';

const useIsomorphicLayoutEffect = typeof window !== 'undefined' ? useLayoutEffect : useEffect;

export const useCurrentHeight = () => {
  const [height, setHeight] = useState(0);

  useIsomorphicLayoutEffect(() => {
    const handleResize = () => {
      if (typeof window === 'undefined') return;
      setHeight(window.innerHeight);
    };

    handleResize();
    window.addEventListener('resize', handleResize);

    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  return height;
};
