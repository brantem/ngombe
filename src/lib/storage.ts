import { openDB, IDBPDatabase, StoreNames, StoreValue, StoreKey } from 'idb';

import { Schema } from 'types/storage';

/* c8 ignore start */
class Storage {
  db: IDBPDatabase<Schema> | undefined;

  _getDB() {
    return openDB('ngombe', 1, {
      async upgrade(db) {
        db.createObjectStore('goals', { keyPath: 'timestamp' });

        db.createObjectStore('records', { keyPath: 'timestamp' });
      },
    });
  }

  async get<Name extends StoreNames<Schema>>(name: Name, key: StoreKey<Schema, Name>) {
    const db = await this._getDB();
    return db.get(name, key);
  }

  async getAll<Name extends StoreNames<Schema>>(name: Name, query: IDBKeyRange) {
    const db = await this._getDB();
    return db.getAll(name, query);
  }

  async put<Name extends StoreNames<Schema>>(name: Name, value: StoreValue<Schema, Name>) {
    const db = await this._getDB();
    return db.put(name, value);
  }

  async add<Name extends StoreNames<Schema>>(name: Name, value: StoreValue<Schema, Name>) {
    const db = await this._getDB();
    return db.add(name, value);
  }

  async delete<Name extends StoreNames<Schema>>(name: Name, key: StoreKey<Schema, Name>) {
    const db = await this._getDB();
    return db.delete(name, key);
  }
}

const storage = new Storage();
export default storage;
/* c8 ignore stop */
