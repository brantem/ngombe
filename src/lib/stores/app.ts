import { create } from 'zustand';

import { goalStore, recordsStore } from 'lib/stores';

type AppState = {
  items: Record<string, any>;
  setItem: (key: string, value: any) => void;
  deleteItem: (key: string) => void;

  date: string | undefined; // YYYY-MM-DD
  setDate(date: string | undefined): void;
};

export const useAppStore = create<AppState>()((set) => ({
  items: {},
  setItem(key, value) {
    set((state) => ({ items: { ...state.items, [key]: value } }));
  },
  deleteItem(key) {
    set((state) => {
      const items = state.items;
      delete items[key];
      return { items };
    });
  },

  date: undefined,
  setDate(date) {
    set({ date });
    goalStore.getState().fetch(date);
    recordsStore.getState().fetch(date);
  },
}));
