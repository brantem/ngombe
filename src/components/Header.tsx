import { cn } from 'lib/helpers';
import { useModal } from 'lib/hooks';
import { useGoalStore } from 'lib/stores';

type HistoriesButtonProps = {
  isBackground?: boolean;
};

const HistoriesButton = ({ isBackground }: HistoriesButtonProps) => {
  const modal = useModal('histories');

  return (
    <button
      className={cn(
        'h-9 px-3 rounded-full border shadow-sm',
        isBackground
          ? 'bg-neutral-100 border-neutral-200 text-neutral-500'
          : 'bg-white border-neutral-100 text-cyan-500',
      )}
      onClick={modal.onOpen}
    >
      Histories
    </button>
  );
};

type GoalButtonProps = {
  isBackground?: boolean;
};

const GoalButton = ({ isBackground }: GoalButtonProps) => {
  const modal = useModal('goal');
  const goal = useGoalStore((state) => state.goal);

  return (
    <button
      className={cn(
        'h-9 px-3 rounded-full border-2 outline-none shadow-sm',
        isBackground ? 'border-neutral-500 text-neutral-500' : 'border-cyan-500 text-cyan-500',
      )}
      onClick={modal.onOpen}
    >
      {goal}ml
    </button>
  );
};

type HeaderProps = {
  isBackground?: boolean;
};

const Header = ({ isBackground }: HeaderProps) => {
  return (
    <header className="w-full px-4 pt-4">
      <div className="max-w-lg w-full mx-auto flex justify-between font-bold">
        <HistoriesButton isBackground={isBackground} />

        <GoalButton isBackground={isBackground} />
      </div>
    </header>
  );
};

export default Header;
