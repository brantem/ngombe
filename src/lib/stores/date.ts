import { create } from 'zustand';

import { goalStore, recordsStore } from 'lib/stores';

type DateState = {
  value: string | undefined;
  set(value: string | undefined): void;
};

// it didn't work when I used useState for this
export const useDateStore = create<DateState>()((set) => ({
  value: undefined,
  set(value) {
    set({ value });
    goalStore.getState().fetch(value);
    recordsStore.getState().fetch(value);
  },
}));
