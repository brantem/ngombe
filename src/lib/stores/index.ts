import { useEffect } from 'react';

import { recordsStore } from 'lib/stores';

/* c8 ignore start */
export const useLoadStore = () => {
  useEffect(() => {
    (async () => {
      recordsStore.getState().setDate(undefined);
    })();
  }, []);
};
/* c8 ignore stop */

export * from 'lib/stores/modal';
export * from 'lib/stores/goal';
export * from 'lib/stores/records';
