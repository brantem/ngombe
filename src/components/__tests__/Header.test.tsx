import { renderHook, act, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import Header from 'components/Header';

import { useAppStore, goalStore } from 'lib/stores';

describe('Header', () => {
  beforeEach(() => {
    act(() => goalStore.setState({ value: 2500 }));
  });

  it('should match snapshot', () => {
    const { container } = render(<Header />);
    expect(container).toMatchSnapshot();
  });

  it('should open records modal', () => {
    const { result } = renderHook(() => useAppStore());
    const setItem = vi.spyOn(result.current, 'setItem');

    render(<Header />);
    act(() => screen.getByTestId('header-records').click());
    expect(setItem).toHaveBeenCalledWith('records', true);
  });

  it('should hide goal button if goal === 0', () => {
    act(() => goalStore.setState({ value: 0 }));

    render(<Header />);
    expect(screen.queryByTestId('header-goal')).not.toBeInTheDocument();
  });

  it('should open goal modal', async () => {
    const { result } = renderHook(() => useAppStore());
    const setItem = vi.spyOn(result.current, 'setItem').mockImplementationOnce(() => {});

    render(<Header />);
    act(() => screen.getByTestId('header-goal').click());
    expect(setItem).toHaveBeenCalledWith('current-goal', true);
  });
});
