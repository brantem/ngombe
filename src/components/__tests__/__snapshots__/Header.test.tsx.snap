// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > should match snapshot 1`] = `
<div>
  <header
    class="w-full px-4 pt-4"
  >
    <div
      class="max-w-lg w-full mx-auto flex justify-between font-bold"
    >
      <button
        class="h-9 px-3 rounded-full border shadow-sm bg-white border-neutral-100 text-cyan-500"
      >
        Histories
      </button>
      <button
        class="h-9 px-3 rounded-full border-2 outline-none shadow-sm border-cyan-500 text-cyan-500"
      >
        2500
        ml
      </button>
    </div>
  </header>
</div>
`;
